### PR TITLE
Preserve scroll position on Android Activity restoration

### PR DESCRIPTION
**Pull request by veyndan**
This doesn't work yet, as `rememberSaveable` needs to support saving of values that _aren't_ `MutableState`'s. Currently, when trying to save the state, we get the following error:

```
app.cash.zipline.ZiplineException: IllegalStateException: unexpected type: {-exvejj=[MutableState(value=medal)@632700585], udeaau=[0]}
                 	at captureStack (runtime/coreRuntime.kt:12)
                 	at IllegalStateException_init_$Create$_0 (kotlin-kotlin-stdlib-js-ir.js)
                 	at toStateSnapshot (commonMainSources/libraries/stdlib/src/kotlin/collections/Maps.kt:84)
                 	at <anonymous> (redwood-treehouse-guest/src/commonMain/kotlin/app/cash/redwood/treehouse/treehouseCompose.kt:114)
                 	at <anonymous> (redwood-treehouse/src/commonMain/kotlin/app/cash/redwood/treehouse/ZiplineTreehouseUi.kt)
                 	…
```

One trade-off here is that we're just saving the first visible item index, and thus we only restore the first visible item index when we come back. What this means is that if we scroll past 4 items, and the top of the `LazyList` has scrolled half of the fifth item, when we restore the scroll position, we will show the whole of the fifth item instead of half of it. I think this is perfectly okay for now, as the gains we get from restoring the scroll index is ginormous.

### CODE REVIEW

**Comment by jingwei99** [comment]
this PR right now is pending on `StateSnapshot` being able to save types (values) that's not `MutableState`. last time @veyndan and I talked on this (it was a month or so ago? maybe more), Veyndan brought an idea about handling serialization for types in `StateSnapshot` (it's just I forgot the details of the idea 🤦🏻‍♀️), I'm talking with Veyndan to get that part done, and then we can land this PR to make the scroll position work!

**Comment by veyndan** [comment]
The idea we were discussing is summarised in https://github.com/cashapp/redwood/issues/1384. We can either go that route of making arbitrary content serializable, or we can just add serialization support for non-`MutableState` values, and implement that issue another day!

File: redwood-lazylayout-compose/build.gradle [hunk 1]
```diff
@@ -13,6 +13,7 @@ kotlin {
   sourceSets {
     commonMain {
       dependencies {
+        implementation libs.jetbrains.compose.runtime.saveable
```
Comment (JakeWharton) [inline redwood-lazylayout-compose/build.gradle]:
```suggestion
        api libs.jetbrains.compose.runtime.saveable
```
This is exposed in public API

**Comment by jingwei99** [comment]
I tested this change locally, with [the change for view id](https://github.com/cashapp/redwood/commit/57f78b982e1a6f7411b4ee0f1e62157ca34491c2), now scroll position is saved!! ✨

**Review by jingwei99 — APPROVED** [review]

**Comment by veyndan** [comment]
@jingwei99 Did you test this with the Activity screen, or one of the sample apps? I'm just asking because I tried testing it on Emoji Search and it isn't working

**Comment by jingwei99** [comment]
> @jingwei99 Did you test this with the Activity screen, or one of the sample apps? I'm just asking because I tried testing it on Emoji Search and it isn't working

I was testing it on the sample app, and it worked. 

🤔 maybe I didn't have the `Don't keep activities` setting on? (I think that's what it is, my emulator has been weird, and I've been wiping its data often)

**Comment by veyndan** [comment]
Superseded by https://github.com/cashapp/redwood/pull/1599

**Pull request closed without merging**